### ISSUE TITLE
nvdaHelperLocal dll import table hook for oleacc SendMessageTimeoutW not being used

### ISSUE THREAD

**Issue by jcsteh** [nvaccess/nvda#14746]
### Steps to reproduce:
1. Press Windows+r to open the Run dialog.
2. Type `\\1.2.3.4` and press enter.
3. NVDA will stop responding because Explorer stops responding. Press alt+tab to switch apps.

### Actual behavior:
NVDA freezes until Explorer responds again.

### Expected behavior:
NVDA shouldn't freeze; watchdog should kill the unresponsive call when you alt+tab.

#### NVDA installed/portable/running from source:
Installed.

#### NVDA version:
alpha-27892,42dbabcc

#### Windows version:
Windows 11 Version 22H2 (OS Build 22621.1413)
Also reproduced on Windows 11 Version 21H2 (OS Build 22000.1696)

#### Relevant NVDA log snippet

```
ERROR - watchdog._watcher (08:56:19.048) - watchdog (21904):
Core frozen in stack!
INFO - watchdog._watcher (08:56:19.051) - watchdog (21904):
Listing stacks for Python threads:
...
Python stack for thread 13448 (MainThread):
  File "nvda.pyw", line 390, in <module>
  File "core.pyc", line 776, in main
  File "wx\core.pyc", line 2237, in MainLoop
  File "gui\__init__.pyc", line 760, in Notify
  File "core.pyc", line 738, in run
  File "queueHandler.pyc", line 88, in pumpAll
  File "queueHandler.pyc", line 55, in flushQueue
  File "scriptHandler.pyc", line 170, in _queueScriptCallback
  File "keyboardHandler.pyc", line 531, in executeScript
  File "inputCore.pyc", line 199, in executeScript
  File "scriptHandler.pyc", line 216, in executeScript
  File "editableText.pyc", line 187, in script_caret_newLine
  File "editableText.pyc", line 86, in _hasCaretMoved
  File "api.pyc", line 368, in processPendingEvents
  File "IAccessibleHandler\__init__.pyc", line 1075, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 724, in processFocusWinEvent
  File "IAccessibleHandler\__init__.pyc", line 576, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 64, in getNVDAObjectFromEvent
  File "IAccessibleHandler\__init__.pyc", line 349, in accessibleObjectFromEvent
  File "oleacc.pyc", line 265, in AccessibleObjectFromEvent
```

#### WinDBG stack for main thread
```
0:018:x86> ~0 kpn
 # ChildEBP RetAddr
00 00cfd814 7614d892 win32u!NtUserMessageCall+0xc
01 00cfd85c 76190a21 USER32!SendMessageTimeoutWorker+0xc2
02 00cfd87c 6f1291d2 USER32!SendMessageTimeoutW+0x21
03 00cfd8b8 6f129082 OLEACC!NativeIAccessibleFromWindow+0x70
04 00cfd8e8 6f150972 OLEACC!AccessibleObjectFromWindow+0x27
05 00cfd924 6f150f97 OLEACC!AccessibleObjectFromEvent+0x5e
06 00cfd938 706ce7ba OLEACC!EXTERNAL_AccessibleObjectFromEvent+0x27
WARNING: Stack unwind information not available. Following frames may be wrong.
07 00cfd958 706cd10b _ctypes!DllCanUnloadNow+0x628a
08 00cfd994 706c926f _ctypes!DllCanUnloadNow+0x4bdb
09 00cfda54 706c9a5e _ctypes!DllCanUnloadNow+0xd3f
0a 00cfdb54 706c59a1 _ctypes!DllCanUnloadNow+0x152e
0b 00cfdba4 702d817b _ctypes+0x59a1
0c 00cfdbc8 703735be python37!PyObject_FastCallKeywords+0x15b
...
```

#### Analysis
We're calling AccessibleObjectFromEvent. That calls SendMessageTimeoutW (frame 2). However, NVDA installs a dll import table hook in oleacc for SendMessageTimeoutW:
`oleaccHooks->requestFunctionHook("USER32.dll", "SendMessageTimeoutW", fake_SendMessageTimeoutW);`
We don't see fake_SendMessageTimeoutW nor cancellableSendMessageTimeout in the stack! We should see this between frames 2 and 3.
This suggests that these hooks aren't working for some reason. That's pretty concerning, since watchdog depends on these to keep NVDA responsive in quite a few cases.

CC @michaelDCurran.

**Comment by CyrilleB79** [comment]
Are 2022.4 / 2023.1rc2 impacted?

I am frequently experimenting NVDA freezes in Explorer, especially when accessing network disks and/or when my VPN is about to disconnect. I wonder if it is related.

**Comment by jcsteh** [comment]
I'm not sure. I suspect this goes back a long way, probably to some build of Windows rather than a specific build of NVDA.

**Comment by leonardder** [comment]
Could this somehow be related to our switch from Minhook to Detours?

**Comment by amirsol81** [comment]
@CyrilleB79 Same here - especially with my VPN about to be disconnected. In general, when Explorer stops responding, NVDA will also become unresponsive until it is restarted manually.

**Comment by jcsteh** [comment]
> Could this somehow be related to our switch from Minhook to Detours?

While I'm not ruling out anything at this point, I don't think so. Hooking SendMessage and friends is a bit risky, so we use our own dll import table hooks (i.e. separate to MinHook and Detours, AKA import address table or IAT hooks) to only hook the SendMessage functions from the perspective of oleacc.dll and UIAutomationCore.dll. The idea is that functions in those dlls that call SendMessage* will get our hooks, but anything else in our process that calls SendMessage* won't. In contrast, when you use MinHook or Detours, the function gets hooked for all code in the process.

It's possible that oleacc and UIAutomationCore now delay load user32.dll, in which case import table hooking won't work. Or maybe user32 functions go via one of those weird api-ms-blah.dll files now, but that's not showing up in the debugger for some reason. Or maybe they're using GetProcAddress directly, though I don't know why they'd do that. Or maybe our dll import table code just broke for some reason.

Probably the simplest fix is just to hook these functions with Detours, though it's a bit risky as noted above.

**Comment by jcsteh** [comment]
Actually, doing this with Detours is going to be tricky, since all the code for dealing with this SendMessage stuff is in nvdaHelperLocal, but the Detours hooking is all in nvdaHelperRemote. Urgh.

**Comment by leonardder** [comment]
Well, after some debugging, it turns out that NVDA never hooks the functions that ought to be hooked. I simply changed some debug logging statements to error for easier debugging and it looks like user32 isn't in the import table for both oleacc and UIAutomationCore. That's really odd though. To be continued.

**Comment by leonardder** [comment]
<details><summary>Log output</summary>
ERROR - NVDAHelperLocal (09:28:35.187) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 31:
_hookFunctionsOnModule: called for OLEACC.dll


ERROR - NVDAHelperLocal (09:28:35.214) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-crt-string-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.229) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-crt-runtime-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.248) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-crt-private-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.260) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for ntdll.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.276) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-sysinfo-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.291) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-libraryloader-l1-2-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.307) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for RPCRT4.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.323) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-string-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.339) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-string-l2-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.354) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-errorhandling-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.370) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-synch-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.385) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-registry-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.400) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-processthreads-l1-1-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.415) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-memory-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.433) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-handle-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.447) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-processthreads-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.462) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-heap-l2-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.478) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-localization-l1-2-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.495) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-synch-l1-2-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.509) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-security-base-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.525) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-debug-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.540) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-libraryloader-l1-2-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.555) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-psapi-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.570) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-profile-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.586) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-interlocked-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.601) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-threadpool-legacy-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.617) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-string-obsolete-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.636) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-heap-obsolete-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.648) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-atoms-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.665) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-delayload-l1-1-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.680) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for OLEACC.dll: module section for api-ms-win-core-delayload-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.727) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 31:
_hookFunctionsOnModule: called for UIAutomationCore.dll


ERROR - NVDAHelperLocal (09:28:35.741) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for msvcp_win.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.759) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-crt-string-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.774) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-crt-runtime-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.790) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-crt-private-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.803) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-sidebyside-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.819) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-libraryloader-l1-2-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.835) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-synch-l1-2-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.850) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-memory-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.866) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-registry-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.881) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-synch-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.899) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-heap-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.913) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-errorhandling-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.931) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-processthreads-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.943) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-eventing-provider-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.960) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-eventing-classicprovider-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.974) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-threadpool-l1-2-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:35.989) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-string-l2-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.008) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-string-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.038) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-localization-l1-2-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.051) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-debug-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.070) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-handle-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.098) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-sysinfo-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.115) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-psapi-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.131) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-libraryloader-l1-2-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.162) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-processthreads-l1-1-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.188) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-profile-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.202) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-interlocked-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.222) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-string-obsolete-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.233) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for ntdll.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.252) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-crt-math-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.266) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-atoms-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.280) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-delayload-l1-1-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.296) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-delayload-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.312) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-synch-l1-2-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.328) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-security-base-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.343) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-file-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.357) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-namedpipe-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.375) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-realtime-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.388) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-heap-l2-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.404) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-io-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.419) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-io-l1-1-1.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.435) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-processthreads-l1-1-3.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.450) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-util-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.469) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-security-appcontainer-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.481) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-threadpool-legacy-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.501) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-kernel32-legacy-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.512) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-shlwapi-legacy-l1-1-0.dll: no requested hooks


ERROR - NVDAHelperLocal (09:28:36.533) - MainThread (12472):
Thread 12472, build\x86\local\dllImportTableHooks.cpp, DllImportTableHooks::hookFunctions, 61:
_hookFunctionsOnModule: import table for UIAutomationCore.dll: module section for api-ms-win-core-apiquery-l1-1-0.dll: no requested hooks
</details>

Long story short, user32 isn't in the import table indeed. It is either one of these api-ms- dlls as @jcsteh  was already afraid of, or even dynamically loaded at runtime as he also suggested.

**Comment by leonardder** [comment]
Als another alternative, we could move apihook.cpp and apihook.h to the common directory and include detours in NVDAHelper_Local as well. What do you think about that?

**Comment by jcsteh** [comment]
Yeah, I considered that. Two reasons not to do that, but they might not be good reasons:

1. I'm not sure if Detours has any process wide state that might cause us trouble if we have two instances running, even if they're in separate dlls. It probably doesn't, but hooking is complex, so... 🤷‍♂️
2. It seems wasteful to duplicate Detours when we already have it elsewhere in our process. It's probably a drop in the ocean in the grand scheme of things, though.

If we did duplicate Detours, we should probably move the clipboard hook out of nvdaHelperRemote as well.

**Comment by leonardder** [comment]
> one possibility might be to move fake_SendMessage* into nvdaHelperRemote, LoadLibrary nvdaHelperLocal.dll in nvdaHelperRemote, use GetProcAddress to get cancellableSendMessageTimeout and then call that in the SendMessage* hook functions. 

I'll try to prototype this later this week.

**Comment by leonardder** [comment]
Ugh, of course. Both cancellableSendMessageTimeout and bgSendMessageThreadProc call SendMessageTimeoutW, so if we going to hook SendMessageTimeoutW in the remote lib, we'll get recursion issues. That's not what we want. I guess integrating detours in the local lib is better after all.
